File: lf-svd.py
import scipy.sparse as sc
import numpy as np

def createA(Y, alpha):
    n = Y.shape[0]
    m = Y.shape[1]
    A = np.zeros((n,m))
    sums_i = np.sum(Y, axis=1)
    sums_j = np.sum(Y, axis=0)
    sum_all = np.sum(Y)

    for i in range(n):
        for j in range(m):
            v = [Y[i,j], 1/n*sums_j[0,j], 1/m*sums_i[i,0], 1/(n*m)*sum_all]
            A[i][j] = np.dot(v, alpha)
            #print(A)
    return A


def createK(A, k):
    # U, S, V = svd(A)
    U, S, V = np.linalg.svd(A, full_matrices=False)
    counter = 0
    for eig in S:
        if eig > k:
            S[counter] = 0
        counter += 1   
    E = np.diag(S)
    nA = U @ E @ V
    return nA

#Load here the matrix with whatever name we provide 
sp_matrix = sc.load_npz("tryout3.npz").todense()
print("printing sp_matrix")
print(sp_matrix)
alpha = [0, 1, 0, 0]
#M = [[1, 0, 0, 0, 1, 0],[0, 1, 1, 0, 0, 0],[0, 0, 1, 1, 1, 1]]
print("starting createA()")
A = createA(sp_matrix,alpha)
print("finished createA()")
print("starting createK()")
nA = createK(A, 1)
print("\n", nA)
print("finished createK(), saving now")
# sc.save_npz("lf-svd-output-from-tryout3.npz", nA, compressed=False)
np.save("lf-svd-output-from-tryout3.npy", nA)
loaded = np.load("lf-svd-output-from-tryout3.npy")
print(loaded)

File: LF-SVD.jl
using EcologicalNetworks
using LinearAlgebra
using NPZ
using PyCall

# ------------- LF Step --------------

# Alpha values:
# a1 -> relative importance of the original value of interaction
# a2 -> in degree
# a3 -> out degree
# a4 -> connectance degree
# Sum of alpha values equals 1


# #Imagine we get adjacency matrix Y
# print("hello")
np = pyimport("numpy")
# pyfile = npzread("tryout.npy.npz")
data = npzread("tryout3.npz")
# #data = np.load(pyfile, allow_pickle = true)
# print(data)


function createA(Y::Matrix, alpha::Vector)
    n = size(Y,1)
    m = size(Y,2)
    A = zeros(n,m)
    sums_j = sum(Y, dims=1)
    print(sums_j)
    sums_i = sum(Y, dims=2)
    print(sums_i)
    sum_all = sum(Y)
    for i=1:n
        for j=1:m
            v = [Y[i,j], 1/n*sums_j[j], 1/m*sums_i[i], 1/(n*m)*sum_all]
            A[i,j] = dot(v, alpha)
            #print(A)
        end
    end
    return A
end

#Start with k=n (rank of the matrix) -> it should be the original matrix
function createK(A::Matrix, k::Int)
    U, S, V = svd(A)
    print(S)
    counter = 1
    for eig in S
        if eig > k
            S[counter] = 0
        end
        counter += 1
    end
    print(S)
    E = Diagonal(S)
    nA = U*E*transpose(V)
    return nA
end
#The value of Aij is the score for the interaction

M = [1 0 0 0 1 0
    0 1 1 0 0 0
    0 0 1 1 1 1]
T = ["A", "B", "C", "D", "E", "F"]
B = ["1", "2", "3"]

# net = AbstractBipartiteNetwork(M, T, B)
# print(net)
alpha = zeros(4)
alpha[2] = 1
A = createA(M, alpha)
nA = createK(A, 1)
